enhancements/network/user-defined-network-segmentation.md

```diff
@@ -0,0 +1,956 @@
+            "interface": "eth1",
+            "ips": [
+                "10.128.0.4"
+            ]},
+            {
+            "mac": "0a:58:0a:f4:02:03",
+            "default": true,
+            "dns": {}
+        }]
+status:
+  phase: Running
+  podIP: 10.244.0.4
+  podIPs:
+  - ip: 10.244.0.4
+```
+
+Multiple namespaces may also be configured to use the same network. In this case the underlying OVN network will be the
+same, following a similar pattern to what is [already supported today for secondary networks](https://docs.openshift.com/container-platform/4.15/networking/multiple_networks/configuring-additional-network.html#configuration-ovnk-network-plugin-json-object_configuring-additional-network).
+
+### CRDs for Managing Networks
+
+Network Attachment Definitions (NADs) are the current way to configure the network in OVN-Kubernetes today, and the
+method proposed in this enhancement. There are two major shortcomings of NAD:
+
+1. It has free-form configuration that depends on the CNI. There is no API validation of what a user enters, leading to
+mistakes which are not caught at configuration time and may cause unexpected functional behavior at runtime.
+2. It requires cluster admin RBAC in order to create the NAD.
+
+In order to address these issues, a proper CRD may be implemented which indirectly creates the NAD for OVN-Kubernetes.
+This solution may consist of more than one CRD, namely an Admin based CRD and one that is namespace scoped for tenants.
+The reasoning behind this is we want tenants to be able to create their own user-defined network for their namespace,
+but we do not want them to be able to connect to another namespace’s network without permission. The Admin based version
+would give higher level access and allow an administrator to create a network that multiple namespaces could connect to.
+It may also expose more settings in the future for networks that would not be safe in the hands of a tenant, like
+deciding if a network is able to reach other services in other networks. With tenants having access to be able to create
+multiple networks, we need to consider potential attack vectors like a tenant trying to exhaust OVN-Kubernetes
+resources by creating too many secondary networks.
+
+Furthermore, by utilizing a CRD, the status of the network CR itself can be used to indicate whether it is configured
+by OVN-Kubernetes. For example, if a user creates a network CR and there is some problem (like pods already existed) then
+an error status can be reported to the CR, rather than relying on the user to check OVN-Kubernetes logs.
+
+### IP Addressing
+
+As previously mentioned, one of the goals is to allow user-defined networks to have overlapping pod IP addresses. This
+is enabled by allowing a user to configure what CIDR to use for pod addressing when they create the network. However,
+this range cannot overlap with the default cluster CIDR used by the cluster default network today.
```

Comment:
Is "I want to _avoid_ having overlapping CIDRs between different user-defined networks" a use case? Because we're doing nothing to support that...

Reply:
I don't see a use case for that yet. Perhaps with BGP and leaking networks into the default VRF that might be requested. For now I think we can defer it until later.